### PR TITLE
DrivAerML: Learnable Fourier Encoding Frequencies

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The DM champion uses fixed Fourier encoding frequencies **[0.5, 2.0, 8.0, 32.0]** for positional encoding of xyz coordinates. These were chosen as reasonable defaults spanning ~2 orders of magnitude. But DrivAerML car surfaces may have preferred spatial scales that don't align with these fixed frequencies.

By making the 4 frequency values learnable parameters (initialized to [0.5, 2.0, 8.0, 32.0]), the model can adapt them during training to the spatial scales most informative for surface field prediction. This is a 4-parameter change — minimal complexity, potentially high impact if the fixed frequencies are suboptimal.

**Key differences from prior Fourier experiments:**
- Fourier NORMALS (brook #3217, dead): added new Fourier features from normal vectors — normals don't benefit from lifting
- This: changes the EXISTING coordinate encoding frequencies from fixed to learnable

---

## Baseline

Current best DrivAerML (PR #3072):
- `val_primary/surface_rel_l2_pct`: **3.833%** at epoch 511
- `test`: 4.685%
- W&B run: `ncl1dh88`
- Config: AdamW lr=5e-4, gc=0.5, no WD, 4L/512d/8H, EMA=0.9995, T_max=30, Fourier freqs=[0.5,2.0,8.0,32.0]

External target: <3.71% (AB-UPT). Gap: ~0.123 pp.

---

## Known Bug Fix

There's a `primary_metric_key` shadowing bug in train.py where a local variable (line ~1646–1648) shadows the function (line ~1428), causing an `UnboundLocalError`. Fix by renaming the local variable to `best_tracking_metric_key` if you encounter it.

---

## Implementation Guidance

1. Find the Fourier frequency definition in the model code (likely in the `FourierFeatures` or similar class).
2. Replace the fixed frequency tensor with `nn.Parameter(torch.tensor([0.5, 2.0, 8.0, 32.0]))` — learnable, initialized to current values.
3. Log the learned frequencies to W&B every 50 epochs (so we can see how they evolve).
4. Use the same learning rate as the rest of the model (no separate LR needed for 4 params).

---

## Trial 1 — Learnable frequencies from default init

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --learnable-fourier-freqs \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --wandb_group dm-learnable-fourier
```

Frequencies initialized to: `[0.5, 2.0, 8.0, 32.0]` (baseline values).

---

## Trial 2 — Learnable frequencies from wider init

Same as Trial 1 but initialize frequencies to `[0.1, 1.0, 10.0, 100.0]` (broader range, let the model compress if needed). Add `--fourier-freq-init "0.1,1.0,10.0,100.0"`.

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --learnable-fourier-freqs \
  --fourier-freq-init "0.1,1.0,10.0,100.0" \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --wandb_group dm-learnable-fourier
```

---

## Key Measurement

Log the **final learned frequencies** for both trials — these tell us which spatial scales the model finds most informative for DrivAerML. This is diagnostically valuable even if the experiment doesn't beat baseline.

---

## Target Metric

Report `val_primary/surface_rel_l2_pct` at best epoch. Target: beat **3.833%**.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT casca:

## Results

### Trial 1 — Default init [0.5, 2.0, 8.0, 32.0]

| Metric | Baseline | Trial 1 | Delta |
|--------|----------|---------|-------|
| `val_primary/surface_rel_l2_pct` | **3.833%** (ep 511) | **5.262%** (ep 179) | +37.3% ❌ |
| `test_primary/surface_rel_l2_pct` | **4.685%** | **6.164%** | +31.6% ❌ |
| External target | <3.71% (AB-UPT) | — | — |

- **Terminal learned frequencies:** [0.317, 2.157, 7.785, 31.818]
- **Diverged at epoch 185** — grad norms spiked from ~0.3 → Infinity. Model never recovered.
- W&B run: `wss8o89v` (group: `dm-learnable-fourier`)
- Peak VRAM: ~12 GB (H100)
- Command:
  ```
  cd target/icml2026 && python train.py \
    --dataset drivaerml --optimizer adamw --lr 5e-4 \
    --cosine-t-max 30 --grad-clip 0.5 --enable-fourier \
    --learnable-fourier-freqs \
    --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
    --epochs 999 --batch-size 1 \
    --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
    --max-train-batches 394 --max-eval-batches 200 \
    --ema-decay 0.9995 --wandb-group dm-learnable-fourier \
    --wandb-name "casca/dm-learnable-fourier-default-init-r2" --agent casca
  ```

### Trial 2 — Wide init [0.1, 1.0, 10.0, 100.0]

| Metric | Baseline | Trial 2 | Delta |
|--------|----------|---------|-------|
| `val_primary/surface_rel_l2_pct` | **3.833%** (ep 511) | **4.252%** (ep 371) | +10.9% ❌ |
| `test_primary/surface_rel_l2_pct` | **4.685%** | **4.961%** | +5.9% ❌ |
| External target | <3.71% (AB-UPT) | — | — |

- **Terminal learned frequencies:** [0.050, 1.528, 10.031, 99.729]
- **Diverged at epoch 429** — grad norms spiked from ~0.5 → 48 → 185 → 1103 → 3524 → Infinity. Model never recovered.
- Experienced multiple earlier perturbation events (eps ~257, 305, 338, 348, 373, 399, 410, 417, 421) that recovered, before the fatal episode at ep 429.
- W&B run: `6cg34qtr` (group: `dm-learnable-fourier`)
- Peak VRAM: ~12 GB (H100)
- Command:
  ```
  cd target/icml2026 && python train.py \
    --dataset drivaerml --optimizer adamw --lr 5e-4 \
    --cosine-t-max 30 --grad-clip 0.5 --enable-fourier \
    --learnable-fourier-freqs --fourier-freq-init "0.1,1.0,10.0,100.0" \
    --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
    --epochs 999 --batch-size 1 \
    --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
    --max-train-batches 394 --max-eval-batches 200 \
    --ema-decay 0.9995 --wandb-group dm-learnable-fourier \
    --wandb-name "casca/dm-learnable-fourier-wide-init-r2" --agent casca
  ```

### What happened

**Both trials diverged catastrophically. The hypothesis is dead.**

Direct parameterization of Fourier frequencies with the same learning rate as model parameters is fundamentally unstable for this setup. Key findings:

1. **Instability mechanism:** Periodic perturbations correlated with cosine LR restart cycles (T_max=30). Each restart injects a sudden LR spike that perturbs the frequencies, which then cascades through all downstream Fourier features. Over time, these perturbations compound and eventually cause irrecoverable divergence.

2. **Init matters for stability but not outcome:** Wide init survived 429 epochs vs 185 for default init — 2.3× more stable — but neither came close to the 3.833% baseline. Wide init's best (4.252%) is still 10.9% worse than baseline.

3. **Frequencies barely moved from init:** Terminal values show the optimizer didn't find dramatically different frequency scales — Trial 2 moved from [0.1, 1.0, 10.0, 100.0] to [0.050, 1.528, 10.031, 99.729], suggesting the original fixed frequencies [0.5, 2.0, 8.0, 32.0] are already reasonable and the model doesn't strongly prefer alternatives.

4. **Even pre-divergence performance was worse:** Before divergence, neither trial tracked toward beating the baseline. The learnable frequencies add optimization difficulty without compensating improvement.

### Suggested follow-ups

If the idea of adaptive frequencies is worth revisiting:
- **Log-space parameterization**: `freq = exp(log_freq_param)` for better gradient scaling across orders of magnitude
- **Separate (much lower) LR** for frequency params: 10–100× lower than model LR to prevent instability
- **Frequency warmup**: freeze freqs for N epochs, then gradually unfreeze
- **Discrete frequency bank + learned attention**: select from a fixed bank of frequencies via soft attention — avoids gradient-through-sinusoid instability entirely

However, finding 3 suggests the fixed frequencies may already be near-optimal, making this direction low-ROI.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3260 because ADVISOR: Closing — dead end. Trial 1 (default init): 5.262% (+37%), Trial 2 (wide init): 4.252% (+11%), both diverged catastrophically post-cosine-restart. Critical diagnostic insight from Trial 2: frequencies barely moved from initialization (terminal [0.05, 1.53, 10.03, 99.73] vs init [0.1, 1.0, 10.0, 100.0]). This strongly suggests the fixed baseline frequencies [0.5, 2.0, 8.0, 32.0] are already well-calibrated for DrivAerML's spatial scales. The instability mechanism is clear: learnable freq + cosine LR restarts = cascading perturbation feedback loop. No variant is worth pursuing since the frequencies don't want to move. Good experimental work..
